File: 40.combination-sum-ii.py
#
# @lc app=leetcode id=40 lang=python3
#
# [40] Combination Sum II
#

# @lc code=start
class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        candidates.sort()
        res = []
        self.checkCombos(candidates, target, 0, 0, [], res)
        return res

    def checkCombos(self, candidates, target, current_idx, current_sum, current_combo, res):
        if current_sum == target:
            res.append(current_combo[:])
            return
        elif current_sum > target:
            return
        
        for i in range(current_idx, len(candidates)):
            if i > current_idx and candidates[i] == candidates[i - 1]:
                continue
            current_combo.append(candidates[i])
            current_sum += candidates[i]
            
            self.checkCombos(candidates, target, i + 1, current_sum, current_combo, res)
            current_combo.pop()
            current_sum -= candidates[i]
        return
        
# @lc code=end

File: 40.combination-sum-ii.java
/*
 * @lc app=leetcode id=40 lang=java
 *
 * [40] Combination Sum II
 */

// @lc code=start

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

class Solution {
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        Arrays.sort(candidates);
        List<List<Integer>> res = new ArrayList<>();
        checkCombos(candidates, target, 0, target, new ArrayList<>(), res);
        return res;
    }

    private void checkCombos(int[] candidates, int target, int currentIdx, int remain, List<Integer> currentCombo, List<List<Integer>> res) {
        
        if (remain == 0) {
            res.add(new ArrayList<>(currentCombo));
            return;
        } else if (remain < 0) {
            return;
        }

        for (int i = currentIdx; i < candidates.length; i++) {
            if (i > currentIdx && candidates[i] == candidates[i - 1]) {
                continue;
            }

            currentCombo.add(candidates[i]);
            checkCombos(candidates, target, i + 1, remain - candidates[i], currentCombo, res);
            currentCombo.remove(currentCombo.size() - 1);
        }
    }
}
// @lc code=end

